### PR TITLE
pragmatically and precisely compute ellipse rectangle on demand

### PR DESCRIPTION
**Pull request by likangning93**
In place of https://github.com/AnalyticalGraphicsInc/cesium/pull/6376.
Fixes https://github.com/AnalyticalGraphicsInc/cesium/issues/6347.

This does the thing where ellipses crossing the poles get circles as well.

Sandcastle for comparison on [cesiumjs](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/#c=tVZtb9s2EP4rhL9YHlzKGRBgSBxjjue1QR0EcLIOBQwUlHSxWVOkQFJO3MH/fUe9S7FbF8P8xTre+8PnTtoxTXYcXkCTGyLhhczA8DSmn7Izrx9m4kxJy7gE3R9cr+RK7tArZNqiksnrXNbKMsuVxDgjOioODcT8nkulp6/coOZy5H5NJftaKS9GTa2r6l5FIFDzz0oS/FUprqrHYa5ph7rqyH4WeiUPWfFFh1upwq1KLbWahVuvSpd3+JzKMEsVsy3MheCJAS8EaUEPynJy3CgecsvBUA2x2sFUCC+PccyGRZFXuLtfogx3aa5IHntYqyBPis007KteK0yv2vLwiO0pXDq2DXQLiO6Z3VCrlizCazZeaTDoOMYMK+dM1I4zJZSmy/kf9IXbzVQkG+aN6GXXcQN8vbHoNqrPD/njoYbQ98nX1Fi01kCelSZ2A0RDaJlcCyhgRsIUgL0HFYPVe6RNk9DztrZ5CTn05MgdnAv1uTCfD/EbFFyHaw0gl2Xn2OAP2FWB9JZEoVI64hJvztXbgY5Wjmfe9Of5YvHw97mX3VHgDArcLaixOoWhu+/CNnbXHgDCabOLLy2tIhE3iWD7o5Gymrol3i6ms4/HaXYoV5pVSgTMbcJIhWmMsNI12LkA93i7v4u8fmGS7cHuImFJIva3XCKua1MvlGEZd3B0/WCGh8CA3rFAQNOrX5KhP6AmDUyoeQBeXne5njxk+CcmUhg0L7ixihOmDfwpFLO16XVt2dpu5UIvDRCXny25Rfyfrrv7QjhWPPklW+bXp4b05s3u/7Vhy5+JV3QTwTNyJWq0PegOyXfRabKoBCp7DzrqfmAyEt136mPo5vcxYSHMd8iowsgrptjg+gEaMoxiXLhmIIojcCeT1E5z+Coc3RvHsbOqvaqzXg8hi0EzmvBwm3ejeFQ50vIVNCStOtZCBUChtC+7dgDWUDQB+xGVDsNTODztE6DLu/cfnr7MFnezjwWWeTXflIqflNfZdLlJb9gbG7sXMMmz/M7jRGlLUi08Sn0LMe4ILMQP0nALloYmcxz7pdM44jvCo5tVr/Ods+qRUDBjUPOcCvHIv8GqNxn7aN9yQ266aX/YgXbbCE02F5NFfkgpHfsovvUq9gGa53WPrRukydgGKtpPakjHVuNhNFkW8zz2UWiqm1J2wh1HiEVAMY3GFY5lk5hLlN5d/IbfZU5kryiWkrGQODETImbZuwA3GJ7s3Lg1PrRIdvBXgjZ43M8y9asWjldg4dW6HAgfSpffz9CKdaTVR5xrkg02ySb7NVsE/wmTJiDu07SJyOUpRNov9/8NllaaU9jgI1LEz4mD/xmPHMNzyv0L).

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Signed CLA is on file.

@likangning93, thanks for the pull request! Maintainers, we have a signed CLA from @likangning93, so you can review this at any time.

:warning: I noticed that [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md) has not been updated. If this change updates the public API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and comment on this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

File: Source/Core/EllipseGeometry.js [hunk 1]
```diff
@@ -805,7 +798,7 @@ define([
         VertexFormat.pack(value._vertexFormat, array, startingIndex);
         startingIndex += VertexFormat.packedLength;
 
-        Rectangle.pack(value._rectangle, array, startingIndex);
+        Rectangle.pack(value.rectangle, array, startingIndex);
```
Comment (hpinkos) [inline Source/Core/EllipseGeometry.js]: Not sure we need to pack the rectangle.  It looks like we're not unpacking it

File: Source/Core/EllipseGeometry.js [hunk 2]
```diff
@@ -980,6 +975,9 @@ define([
          */
         rectangle : {
             get : function() {
+                if (!defined(this._rectangle)) {
+                    this._rectangle = computeRectangle(this);
```
Comment (hpinkos) [inline Source/Core/EllipseGeometry.js]: It can be in a separate PR, but we should also update `polygon`, `corridor` and `rectangle` to work this way

**Comment by hpinkos** [comment]
Looks good @likangning93!  Those are my only comments

**Comment by likangning93** [comment]
@hpinkos this is up-to-date and passing for me locally, but I can't see CI for some reason?

**Comment by hpinkos** [comment]
Yeah I've seen travis do that before where it just totally misses running on a commit

File: Source/Core/CircleGeometry.js [hunk 3]
```diff
@@ -15,6 +16,7 @@ define([
         defineProperties,
         EllipseGeometry,
         Ellipsoid,
+        Rectangle,
```
Comment (hpinkos) [inline Source/Core/CircleGeometry.js]: Is this needed anymore?

**Comment by hpinkos** [comment]
I think this warrants an entry in `CHANGES.md` too

**Comment by likangning93** [comment]
> I think this warrants an entry in CHANGES.md too

geometry rectangles are currently private in the API though?

**Comment by hpinkos** [comment]
> geometry rectangles are currently private in the API though

Oh right, I forgot about that.  Nevermind =)

**Comment by likangning93** [comment]
@hpinkos Travis is done, this is ready!

**Comment by hpinkos** [comment]
Awesome job @likangning93!  Gold star for the sandcastle example, that made testing this really easy.
I'll work on adding the rectangle changes to the other ground geometry types.

**Comment by pjcozzi** [comment]
Update CHANGES.md?

**Comment by hpinkos** [comment]
@pjcozzi no because it's a change to part of the private API

**Comment by pjcozzi** [comment]
OK, I thought it fixed a bug.
